vm/systemSmartContracts/auction.go

```diff
@@ -0,0 +1,694 @@
+	NumNodes      uint32   `json:"NumNodes"`
+	TotalSupply   *big.Int `json:"TotalSupply"`
+	MinStep       *big.Int `json:"MinStep"`
+	NodePrice     *big.Int `json:"NodePrice"`
+}
+
+type stakingAuctionSC struct {
+	eei           vm.SystemEI
+	unBoundPeriod uint64
+	kg            crypto.KeyGenerator
+	baseConfig    AuctionConfig
+}
+
+// NewStakingAuctionSmartContract creates an auction smart contract
+func NewStakingAuctionSmartContract(
+	minStakeValue *big.Int,
+	minStepValue *big.Int,
+	totalSupply *big.Int,
+	unBoundPeriod uint64,
+	numNodes uint32,
+	eei vm.SystemEI,
+	kg crypto.KeyGenerator,
+) (*stakingAuctionSC, error) {
+	if minStakeValue == nil {
+		return nil, vm.ErrNilInitialStakeValue
+	}
+	if minStakeValue.Cmp(big.NewInt(0)) < 1 {
+		return nil, vm.ErrNegativeInitialStakeValue
+	}
+	if check.IfNil(eei) {
+		return nil, vm.ErrNilSystemEnvironmentInterface
+	}
+
+	baseConfig := AuctionConfig{
+		MinStakeValue: minStakeValue,
+		NumNodes:      numNodes,
+		TotalSupply:   totalSupply,
+		MinStep:       minStepValue,
+		NodePrice:     big.NewInt(0).Set(minStakeValue),
+	}
+
+	reg := &stakingAuctionSC{
+		eei:           eei,
+		unBoundPeriod: unBoundPeriod,
+		kg:            kg,
+		baseConfig:    baseConfig,
+	}
+	return reg, nil
+}
+
+// Execute calls one of the functions from the staking smart contract and runs the code according to the input
+func (s *stakingAuctionSC) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if CheckIfNil(args) != nil {
+		return vmcommon.UserError
+	}
+
+	switch args.Function {
+	case "_init":
+		return s.init(args)
+	case "stake":
+		return s.stake(args)
+	case "unStake":
+		return s.unStake(args)
+	case "unBound":
+		return s.unBound(args)
+	case "claim":
+		return s.claim(args)
+	case "slash":
+		return s.slash(args)
+	case "get":
+		return s.get(args)
+	}
+
+	return vmcommon.UserError
+}
+
+func (s *stakingAuctionSC) get(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if len(args.Arguments) < 1 {
+		return vmcommon.UserError
+	}
+
+	value := s.eei.GetStorage(args.Arguments[0])
+	s.eei.Finish(value)
+
+	return vmcommon.Ok
+}
+
+func (s *stakingAuctionSC) setConfig(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	ownerAddress := s.eei.GetStorage([]byte(ownerKey))
+	if !bytes.Equal(ownerAddress, args.CallerAddr) {
+		log.Debug("setConfig function called by not the owners address")
+		return vmcommon.UserError
+	}
+
+	if len(args.Arguments) != 6 {
+		log.Debug("setConfig function called with wrong number of arguments")
+		return vmcommon.UserError
+	}
+
+	config := AuctionConfig{
+		MinStakeValue: big.NewInt(0).SetBytes(args.Arguments[0]),
+		NumNodes:      uint32(big.NewInt(0).SetBytes(args.Arguments[1]).Uint64()),
+		TotalSupply:   big.NewInt(0).SetBytes(args.Arguments[2]),
+		MinStep:       big.NewInt(0).SetBytes(args.Arguments[3]),
+		NodePrice:     big.NewInt(0).SetBytes(args.Arguments[4]),
+	}
+
+	configData, err := json.Marshal(config)
```

Comment:
i don't know it is worth it.

Reply:
Taking a step back this solution is not as bad as it seemed for me because the data consistency is assured when changing protocol marshalizer and, the most important, this is a concern *only* for the VM implementation. Same thing happens with other VMs like Arwen or iele, they output raw bytes as value because the encoding/decoding is part of the VM implementation. The protocol does not care about value types, it only known []byte and this should remain like this. Choosing json might not be the most optimum solution but it will serve us well especially when debugging.

Reply:
Answered above